netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java

```diff
@@ -182,7 +183,13 @@ public ClientStream newStream(MethodDescriptor<?, ?> method, Metadata.Headers he
     Http2Headers http2Headers = Utils.convertClientHeaders(headers, ssl, defaultPath, authority);
 
     // Write the command requesting the creation of the stream.
-    channel.writeAndFlush(new CreateStreamCommand(http2Headers, stream));
+    CreateStreamCommand msg = new CreateStreamCommand(http2Headers, stream);
+    if (method.getType().clientSendsOneMessage()) {
+      // No need to flush yet if the client is required to send at least one message
+      channel.write(msg);
```

Comment:
If nothing is buffered, the flush will be relatively cheap; basically the cost of the event going through the pipeline.

If the CreateStreamCommand is queued, then it won't need to be marked for future flushing. That's because when it is written it will be written due to some other frame, and that other frame will already need proper flushing. For instance, let's say we receive a frame with END_STREAM in connectionRead() that makes a stream advance to CLOSED. We then write out a new HEADERS frame and maybe even some DATA. But then on connectionReadComplete() we would be flushing, so it is properly flushed without any special magic.